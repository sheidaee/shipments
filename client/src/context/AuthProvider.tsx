import React, { useState } from 'react';
import Cookies from 'js-cookie';
import { default as axios } from "../utilities/axios-conf";

export const AuthContext = React.createContext({});

// const fakeUserData: any = {
//   id: 1,
//   name: 'Jhon Doe',
//   avatar: '',
//   roles: ['USER', 'ADMIN'],
// };

/**
 * We have used Fake JWT token from "jwt.io"
 * This is a sample token to show auth is working
 * Your token will come from your server when user successfully loggedIn.
 */

const fakeToken: any =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjEiLCJuYW1lIjoidGFyZXEgam9iYXllcmUiLCJyb2xlcyI6ImFkbWluIn0.k74_B-zeWket405dIAt018mnQFMh_6_BTFpjB77HtRQ';

const addItem = (key: string, value = '') => {
  /**
   *  Using the local storage code....
   */
  // if (key) localStorage.setItem(key, value);

  /**
   *  Using the Cookies code...
   */
  if (key) Cookies.set(key, value, { expires: 7 });
};

const clearItem = (key: any) => {
  /**
   *  Using the local storage code....
   */
  // localStorage.removeItem(key);

  /**
   *  Using the Cookies code...
   */
  Cookies.remove(key);
};

// const isValidToken = () => {
//   /**
//    *  Using the local storage code....
//    */
//   // const token = localStorage.getItem('token');

//   /**
//    *  Using the Cookies code...
//    */
//   const token = Cookies.get('token');
//   // JWT decode & check token validity & expiration.
//   if (token) return true;
  
//   return false;
// };

const AuthProvider = (props: any) => {
  // const [loggedOut, setLoggedOut] = useState(true);
  const [user, setUser] = useState(null);
  const [loggedIn, setLoggedIn] = useState(!!user); // isValidToken()
  const [token, setToken] = useState(null);
  const [authMessage, setAuthMessage] = useState("");

  const hasRole = (roleName: string) => {
    if (!roleName) {
      return false;
    }

    if (!loggedIn) {
      return false;
    }

    if (!user) {
      return false;
    }

    return (user as any).roles.includes(roleName);
  }
  
  const signIn = async(params: any) => {    
    try {
      setAuthMessage("");

      const { username, password } = params;
      const result = await axios.post("auth/signin", {
        username,
        password
        });

      if (result.status === 200) {
        console.log(params, 'sign in form Props');
        console.log(result.data);
      
        setUser(result.data);
        setToken(fakeToken);
        addItem('token', fakeToken);
        setLoggedIn(true);        
      }
      if (result.status === 400) {
        return result.data.message;
      }
    }
    catch(e) {
      console.log(e);  

      setAuthMessage(e.response.data.message);

      return false;
    };
  };

  const logOut = () => {
    setUser(null);
    setToken(null);
    clearItem('token');
    setLoggedIn(false);
  };

  return (
    <AuthContext.Provider
      value={{
        loggedIn,
        logOut,
        signIn,
        user,
        token,
        authMessage,
        hasRole
      }}
    >
      <>{props.children}</>
    </AuthContext.Provider>
  );
};

export default AuthProvider;
