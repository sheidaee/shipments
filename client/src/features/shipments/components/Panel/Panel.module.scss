.Panel {
  padding: 2rem;
  text-align: center;
  height: 40vh;
  background-image: linear-gradient(rgba(0, 0, 0, 0.35), rgba(0, 0, 0, 0.35)),
    url("/images/back.png");
  background-size: cover;
  background-position: center;
  position: relative;
  color: #fff;
  line-height: 1.7;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
}

.metaData {
  font-size: 1.8rem;
  text-align: center;
  margin-bottom: 1rem;
  font-weight: 300;
  
  span {
    display: inline-block;

    &:not(:last-child) {
      margin-right: 1rem;      
    }
  }
}

.title {
  font-weight: 300;
  font-size: 4.6rem;
  text-align: center;
  margin-bottom: 2.5rem;
  letter-spacing: 0.2rem;
}

.btn {
  margin: 0 2rem;
}
