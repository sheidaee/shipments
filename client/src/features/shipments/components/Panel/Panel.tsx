import React, { useContext } from "react";
import moment from "moment";
import {  
  useHistory, 
  Redirect
} from "react-router-dom";

import { AuthContext } from "../../../../context/AuthProvider";
import AddShipment from "../AddShipment";
import Styles from "./Panel.module.scss";
import SignIn from "../../../Authentication/components/SingIn";
import SignOut from "../../../Authentication/components/SignOut";
import { ADMIN_HOME_PAGE, ASSIGNEE_HOME_PAGE } from "../../../../utilities/constants";
import IconEl from "../../../../components/IconEl";

/**
 * Displays current date, total number of shipments, add shipment button
 *
 */
export function Panel({ shipmentsCount = 0 }) {
  const { loggedIn, hasRole, user }: any = useContext(AuthContext);
  const { location } = useHistory();    
  
  if (location.pathname === "/" && loggedIn) {
    const redirectTo = hasRole('ADMIN') ? ADMIN_HOME_PAGE : ASSIGNEE_HOME_PAGE;
    return <Redirect to={redirectTo} />;
  }
  
  return (
    <div className={Styles.Panel}>
      <div className={Styles.metaData}>
        {loggedIn && <span><IconEl icon="user" /> {user.name} </span>} 
        <span><IconEl icon="calendar" /> { moment().format("MMM Do YY")}</span>
      </div>
      {loggedIn && <>      
          <div className={Styles.title}>Total: {shipmentsCount}</div>        
          {hasRole('ADMIN') && <AddShipment />}
          <SignOut />
        </>
      }

      {!loggedIn && <SignIn />}
    </div>
  );
}

export default Panel;
