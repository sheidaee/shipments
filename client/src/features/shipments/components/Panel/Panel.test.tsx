import React from "react";
import { configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { BrowserRouter as Router } from "react-router-dom";
import moment from "moment";

import { Panel } from "./Panel";
import AuthProvider from "../../../../context/AuthProvider";

configure({ adapter: new Adapter() });

describe("<Panel />", () => {
  let wrapper: any;
  const currentDate = moment().format("MMM Do YY");

  beforeEach(() => {
    wrapper = mount(<Router><AuthProvider><Panel /></AuthProvider></Router>);
  });

  it(`Should display current date ${currentDate}`, () => {
    expect(wrapper.contains(currentDate)).toEqual(true);
  });
});
