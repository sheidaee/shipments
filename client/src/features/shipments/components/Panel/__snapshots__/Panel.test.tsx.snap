// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> renders and matches snapshot 1`] = `
<Panel
  customersCount={1}
>
  <div
    className="Panel"
  >
    <div
      className="date"
    >
      Oct 27th 19
    </div>
    <div
      className="title"
    >
      Total: 
      1
    </div>
    <withState(AddCustomer)>
      <Blueprint3.Button
        className=""
        onClick={[Function]}
        text="Add customer"
      >
        <button
          className="bp3-button"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          type="button"
        >
          <Blueprint3.Icon
            key="leftIcon"
          />
          <span
            className="bp3-button-text"
            key="text"
          >
            Add customer
          </span>
          <Blueprint3.Icon
            key="rightIcon"
          />
        </button>
      </Blueprint3.Button>
      <Blueprint3.Dialog
        canOutsideClickClose={false}
        icon="inbox"
        isOpen={false}
        onClose={[Function]}
        title="Add Customer"
      >
        <Blueprint3.Overlay
          autoFocus={true}
          backdropProps={Object {}}
          canEscapeKeyClose={true}
          canOutsideClickClose={false}
          className="bp3-overlay-scroll-container"
          enforceFocus={true}
          hasBackdrop={true}
          icon="inbox"
          isOpen={false}
          lazy={true}
          onClose={[Function]}
          title="Add Customer"
          transitionDuration={300}
          transitionName="bp3-overlay"
          usePortal={true}
        />
      </Blueprint3.Dialog>
    </withState(AddCustomer)>
  </div>
</Panel>
`;
